File: examples/spectral/dynamical_thermal/mpi_sz_spatial.py
"""spectral | dynamical thermal | MPI (distributed) | U(1)-Sz x cyclic translation

finite-T S_zz(omega, T) at three (T, omega) probe points. Twin: ``examples/spectral/dynamical_thermal/mpi_sz_spatial.cpp``.

Requires: WITH_MPI build + launch via mpirun -n <ranks>
"""
from __future__ import annotations

import os
import sys

import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from _shared.common import heisenberg_chain, rank0_print

import qed

N = 8
H = heisenberg_chain(N, pbc=True)

obs = qed.Operator(N, 0.5)
for i in range(N):
    obs.add_one_body(qed.OP_SZ, i, 1.0)

omega = np.linspace(-5.0, 5.0, 11)
result = qed.spectral(
    H,
    [obs],
    method="ftlm_dynamical",
    T=[0.5, 1.0, 2.0],
    omega=omega,
    eta=0.1,
    krylov_dim=40,
    num_random_vectors=8,
    device="mpi",
    sz=N // 2,
    symmetry=[[(i + 1) % N for i in range(N)]],
    verbose=False,
)

mid = len(result.omega) // 2
rank0_print(f"S(w=-5.0) = {result.S_real[0]:.6f}")
rank0_print(f"S(w= 0.0) = {result.S_real[mid]:.6f}")
rank0_print(f"S(w= 5.0) = {result.S_real[-1]:.6f}")

# === Expected output (deterministic; captured on the CI reference runner) ===
# (filled in by refresh_expected_output.py once the CPU binaries are built)
# ===========================================================================

File: examples/solve/lanczos/mpi_sz_spatial.cpp
// =============================================================================
// examples/solve/lanczos/mpi_sz_spatial.cpp
//
// solve | LANCZOS | MPI (distributed) | U(1)-Sz x cyclic translation (Sz=0)
//
// single ground-state eigenvalue via Lanczos. Twin: examples/solve/lanczos/mpi_sz_spatial.py
//
// Requires: WITH_MPI build + launch via mpirun -n <ranks>
// =============================================================================

#include "../../_shared/common.h"

#include <ed/api.h>
#include <ed/api/symmetry_helpers.h>

#include <cmath>
#include <iomanip>
#include <iostream>

int main(int argc, char** argv) {
    auto guard = ed_example::mpi_guard(argc, argv);
    constexpr std::uint64_t N = 8;
    (void)guard;

    auto op = ed_example::heisenberg_chain(N, /*pbc=*/true);
    op->symmetry_info = ed::find_symmetries(static_cast<int>(N), "translation");

    auto spec = ed_example::in_memory_spec(std::move(op), N);

    ed::api::SolveOptions opts;
    opts.num_eigenvalues = 1;
    opts.solver          = "LANCZOS";
    opts.device          = "mpi";
    opts.sz              = static_cast<int>(N / 2);

    opts.tolerance       = 1e-10;

    auto result = ed::api::solve(std::move(spec), opts);

    std::cout << std::setprecision(10);
    for (std::size_t k = 0; k < result.eigenvalues.size(); ++k) {
        ed_example::rank0_print("E[", k, "] = ", result.eigenvalues[k], "\n");
    }
    const double E0 = result.eigenvalues.front();
    const double Eref = ed_example::bethe_E0(N);
    if (std::isfinite(Eref)) {
        ed_example::rank0_print("|E0 - E0_Bethe| = ",
                                 std::scientific, std::setprecision(2),
                                 std::abs(E0 - Eref), "\n");
    }

    // === Expected output (deterministic; captured on the CI reference runner) ===
    // E[0] = -3.6510934089   (filled in by refresh_expected_output.py)  (captured on the mpi reference runner)
    // |E0 - E0_Bethe| ~ 1e-10
    // ===========================================================================
    return 0;
}
